test/ports_test.dart

```diff
@@ -22,259 +22,335 @@ void main() {
     singleCompletePort(completer,
         timeout: _ms * 100, onTimeout: () => throw 91);
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       fail('unreachable');
     }, onError: (e, s) {
       expect(e, 91);
     });
   });
 
   test('TimeoutCallbackFuture', () {
-    var completer = Completer.sync();
+    final completer = Completer.sync();
     singleCompletePort(completer,
         timeout: _ms * 100, onTimeout: () => Future.value(87));
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       expect(v, 87);
     });
   });
 
   test('TimeoutCallbackThrowsFuture', () {
-    var completer = Completer.sync();
+    final completer = Completer.sync();
     singleCompletePort(completer,
         timeout: _ms * 100, onTimeout: () => Future.error(92));
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       fail('unreachable');
     }, onError: (e, s) {
       expect(e, 92);
     });
   });
 
   test('TimeoutCallbackSLow', () {
-    var completer = Completer.sync();
+    final completer = Completer.sync();
     singleCompletePort(completer,
         timeout: _ms * 100,
         onTimeout: () => Future.delayed(_ms * 500, () => 87));
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       expect(v, 87);
     });
   });
 
   test('TimeoutCallbackThrowsSlow', () {
-    var completer = Completer.sync();
+    final completer = Completer.sync();
     singleCompletePort(completer,
         timeout: _ms * 100,
         onTimeout: () => Future.delayed(_ms * 500, () => throw 87));
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       fail('unreachable');
     }, onError: (e, s) {
       expect(e, 87);
     });
   });
 
   test('TimeoutFirst', () {
-    var completer = Completer.sync();
-    var p =
+    final completer = Completer.sync();
+    final p =
         singleCompletePort(completer, timeout: _ms * 100, onTimeout: () => 37);
     Timer(_ms * 500, () => p.send(42));
-    return completer.future.then((v) {
+    return completer.future.then<Null>((v) {
       expect(v, 37);
     });
   });
+
+  test('TimeoutFirst with valid null', () {
+    final completer = Completer<int?>.sync();
+    final p = singleCompletePort(completer,
+        timeout: _ms * 100, onTimeout: () => null);
+    Timer(_ms * 500, () => p.send(42));
+    return expectLater(completer.future, completion(null));
+  });
+
+  test('TimeoutFirst with invalid null', () {
+    final completer = Completer<int>.sync();
+
+    /// example of incomplete generic parameters promotion.
+    /// same code with [singleCompletePort<int, dynamic>] is a compile time error
+    final p = singleCompletePort(
+      completer,
+      timeout: _ms * 100,
+      onTimeout: () => null,
+    );
+    Timer(_ms * 500, () => p.send(42));
+    return expectLater(completer.future, throwsA(isA<TypeError>()));
+  });
```

Comment:
this seems to be related to generic types promotion with incomplete types
`callback` arg is required to promote second generic parameter, so both parameters are handled as dynamic

Reply:
Seems likely, yes.